hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenAirdropHandler.java

```diff
@@ -252,148 +194,210 @@ final var record =
+                if (currentHeadAirdrop == null) {
+                    log.error(
+                            "Head pending airdrop {} not found for account {}",
+                            currentHeadAirdropId,
+                            senderAccount.accountId());
+                    newHeadAirdrop = createFirstAccountPendingAirdrop(pendingValue);
+                } else {
+                    final var updatedHeadAirdrop = currentHeadAirdrop
+                            .copyBuilder()
+                            .previousAirdrop(pendingId)
+                            .build();
+                    // since we already validated the headAirdropId exists, we can safely update the store
+                    pendingStore.put(currentHeadAirdropId, updatedHeadAirdrop);
+                    // Create new account airdrop with next airdrop ID the previous head airdrop
+                    newHeadAirdrop = createAccountPendingAirdrop(pendingValue, currentHeadAirdropId);
+                }
+            } else {
+                newHeadAirdrop = createFirstAccountPendingAirdrop(pendingValue);
+            }
+            // Update the sender account with new head pending airdrop
+            final var numPendingAirdrops = senderAccount.numberPendingAirdrops();
+            final var updatedSenderAccount = senderAccount
                     .copyBuilder()
-                    .previousAirdrop(pendingId)
+                    .headPendingAirdropId(pendingId)
+                    .numberPendingAirdrops(numPendingAirdrops + 1)
                     .build();
-            pendingStore.patch(headAirdropId, updatedAirdrop);
-
-            // Create new account airdrop with next airdrop ID the previous head airdrop
-            newAccountPendingAirdrop = createAccountPendingAirdrop(pendingValue, headAirdropId);
-        } else {
-            newAccountPendingAirdrop = AirdropHandlerHelper.createAccountPendingAirdrop(pendingValue);
+            accountStore.put(updatedSenderAccount);
+            pendingStore.put(pendingId, newHeadAirdrop);
         }
-        // Update the sender account with new head pending airdrop
-        final var updatedSenderAccount =
-                senderAccount.copyBuilder().headPendingAirdropId(pendingId).build();
-        accountStore.put(updatedSenderAccount);
-        return newAccountPendingAirdrop;
     }
 
     /**
-     *  Calculate the fees for the token airdrop transaction. The fees are calculated by combining the
-     *  default airdrop fees, the crypto transfer fees and the token association fees.
+     * Gets the airdrop fee for the token airdrop transaction.
+     * @param feeContext the fee context
+     * @return the airdrop fee
+     */
+    private long airdropFeeFor(@NonNull final FeeContext feeContext) {
+        return feeContext
+                .feeCalculatorFactory()
+                .feeCalculator(SubType.DEFAULT)
+                .calculate()
+                .totalFee();
```

Comment:
With this code wouldn't we always charge $0.1? Shouldn't we check for associations? If there are we should charge - $0.05, if there aren't - $0.1.

Reply:
Discussed with Ali. It seems these are the fees. Will update the implementation

> Here are 3 possible scenarios, let's assume you are airdropping token A to me:
> - If I explicitly associated with token A, we only charge the CryptoTransfer fee of $0.001, 
> - If I have open slots or unlimited association (maxAutomaticAssociations: -1), meaning I don't have explicit association to token A but can receive airdrops, then TokenAirdrop would cost $0.1
> - If I have no open slots (maxAutomaticAssociations: 0), which means the airdrop will go to pending state, then TokenAirdrop would be $0.1 still.